tests/migrations/test_operations.py

```diff
@@ -1972,6 +2019,99 @@ def test_remove_partial_unique_constraint(self):
             'name': 'test_constraint_pony_pink_for_weight_gt_5_uniq',
         })
 
+    def test_add_deferrable_unique_constraint(self):
+        project_state = self.set_up_test_model('test_adddeferrableuniqueconstraint')
+        deferrable_unique_constraint = models.UniqueConstraint(
+            fields=['pink'],
+            name='test_constraint_pony_pink_deferrable',
+            defer=models.Deferrable.DEFERRED,
+        )
+        operation = migrations.AddConstraint('Pony', deferrable_unique_constraint)
+        self.assertEqual(
+            operation.describe(),
+            'Create constraint test_constraint_pony_pink_deferrable '
+            'on model Pony'
+        )
+        # Test the state alteration
+        new_state = project_state.clone()
+        operation.state_forwards('test_adddeferrableuniqueconstraint', new_state)
+        self.assertEqual(len(new_state.models['test_adddeferrableuniqueconstraint', 'pony'].options['constraints']), 1)
+        Pony = new_state.apps.get_model('test_adddeferrableuniqueconstraint', 'Pony')
+        self.assertEqual(len(Pony._meta.constraints), 1)
+        # Test the database alteration
+        with connection.schema_editor() as editor:
+            operation.database_forwards('test_adddeferrableuniqueconstraint', editor, project_state, new_state)
+        # Test constraint works
+        Pony.objects.create(pink=1, weight=4.0)
+        with atomic():
+            if connection.features.supports_deferrable_unique_constraints:
+                Pony.objects.create(pink=1, weight=4.0)
+                with self.assertRaises(IntegrityError):
```

Comment:
To catch this as a `IntegrityError` on Oracle we need to add `ORA-00001` to the wrapper:
```diff --git a/django/db/backends/oracle/base.py b/django/db/backends/oracle/base.py
index 4d2f5b51f1..3908aebda1 100644
--- a/django/db/backends/oracle/base.py
+++ b/django/db/backends/oracle/base.py
@@ -73,7 +73,7 @@ def wrap_oracle_errors():
         #               _C00102056) violated - parent key not found'
         # Convert that case to Django's IntegrityError exception.
         x = e.args[0]
-        if hasattr(x, 'code') and hasattr(x, 'message') and x.code == 2091 and 'ORA-02291' in x.message:
+        if hasattr(x, 'code') and hasattr(x, 'message') and x.code == 2091 and ('ORA-02291' in x.message or 'ORA-00001' in x.message):
             raise IntegrityError(*tuple(e.args))
         raise
 
```

Reply:
I should update the comment too - I think that needs the actual oracle error message though, and I don't know it.

Reply:
Oh, you included an example above!